### PR TITLE
NEW Add TaxonomyType to group taxonomy terms by different types

### LINKED ISSUE #26: Enhancement: Add "type" to TaxonomyTerm
As a CMS user,
I want to be able to group taxonomy terms into "types",
So that when I add taxonomy terms to a type of object in the CMS I can only show relevant taxonomy terms for my "type"

### Acceptance criteria

* You can define a "type" for a taxonomy term
* Public API remains stable (minor version update)
* A new public API method exists to retrieve taxonomy terms by a given type (e.g. `TaxonmyTerm::getByType`)

### PR DESCRIPTION
**Pull request by robbieaverill**
Resolves #26

### CODE REVIEW

File: code/TaxonomyTerm.php [hunk 1]
```diff
@@ -32,6 +33,11 @@ class TaxonomyTerm extends DataObject implements PermissionProvider
 
     private static $default_sort = 'Sort';
 
+    private static $summary_fields = array(
+        'Name' => 'Name',
+        'getTaxonomyType' => 'Type'
```
Comment (tractorcow) [inline code/TaxonomyTerm.php]: For getters just use the property name `TaxonomyType` not the method of the getter.

File: code/TaxonomyTerm.php [hunk 2]
```diff
@@ -86,15 +97,47 @@ public function getTaxonomyName()
...
     }
 
+    /**
+     * Get the type of the top-level ancestor
+     *
+     * @return string
+     */
+    public function getTaxonomyType()
+    {
+        return $this->getTaxonomy()->Type()->Name;
```
Comment (tractorcow) [inline code/TaxonomyTerm.php]: you need to check for nulls along this chain.

File: code/TaxonomyAdmin.php [hunk 3]
```diff
@@ -1,28 +1,40 @@
...
+        if ($this->modelClass === 'TaxonomyTerm') {
+            $list = parent::getList();
+            return $list->filter('ParentID', '0');
+        }
+        return parent::getList();
     }
 
     public function getEditForm($id = null, $fields = null)
     {
+        if (!$this->modelClass !== 'TaxonomyTerm') {
```
Comment (fspringveldt) [inline code/TaxonomyAdmin.php]: Shouldn't we just say here if($this->modelClass == 'TaxonomyTerm')

Comment (robbieaverill) [inline code/TaxonomyAdmin.php]: The pre-existing logic from this method only applies to `TaxonomyTerm` - adding a second managed model made it do bad things. This is a check to ensure we don't do bad things for `TaxonomyType` models

Comment (tractorcow) [inline code/TaxonomyAdmin.php]: I think @fspringveldt means "use a single positive" instead of "double negative"

Comment (robbieaverill) [inline code/TaxonomyAdmin.php]: I hadn't even noticed that (facepalm)

Comment (robbieaverill) [inline code/TaxonomyAdmin.php]: Updated

File: code/TaxonomyTerm.php [hunk 4]
```diff
@@ -86,15 +97,57 @@ public function getTaxonomyName()
...
 
+    /**
+     * Get the type of the top-level ancestor if it is set
+     *
+     * @return string
+     */
+    public function getTaxonomyType()
+    {
+        $taxonomy = $this->getTaxonomy();
+        if ($taxonomy->Type()->exists()) {
```
Comment (tractorcow) [inline code/TaxonomyTerm.php]: You fixed this by calling ->exists() but calling that on `null` is still an error.

**Pull request merged**